src/trace/trace.cpp

```diff
@@ -308,91 +321,49 @@ void Trace::add_lo2s_property(const std::string& name, const std::string& value)
-        registry_.get<otf2::definition::location_group>(ByCpu(cpuid)),
-        otf2::definition::location::location_type::cpu_thread);
+    MetricScope metric_scope(writer_scope);
 
-    comm_locations_group_.add_member(location);
-    return archive_(location);
-#endif
+    const auto& intern_location = registry_.emplace<otf2::definition::location>(
+        ByMeasurementScope(metric_scope), intern(metric_scope.name()),
+        registry_.get<otf2::definition::location_group>(
+            ByExecutionScope(groups_.get_group(writer_scope))),
+        otf2::definition::location::location_type::metric);
+    return archive_(intern_location);
 }
 
-otf2::writer::local& Trace::cpu_switch_writer(int cpuid)
+otf2::writer::local& Trace::switch_writer(const ExecutionScope& writer_scope)
 {
-    // As the cpuid is unique in this context, create only one writer/location per
-    // cpuid
-    auto name = intern(fmt::format("cpu {}", cpuid));
+    SwitchScope scope(writer_scope);
 
-    const auto& location = registry_.emplace<otf2::definition::location>(
-        ByCpuSwitchWriter(cpuid), name,
-        registry_.get<otf2::definition::location_group>(ByCpu(cpuid)),
+    const auto& intern_location = registry_.emplace<otf2::definition::location>(
+        ByMeasurementScope(scope), intern(scope.name()),
+        registry_.get<otf2::definition::location_group>(
+            ByExecutionScope(groups_.get_group(writer_scope))),
         otf2::definition::location::location_type::cpu_thread);
 
-    comm_locations_group_.add_member(location);
-    return archive_(location);
-}
-
-otf2::writer::local& Trace::thread_metric_writer(pid_t pid, pid_t tid)
-{
-    auto name = fmt::format("metrics for thread {}", tid);
+    comm_locations_group_.add_member(intern_location);
 
-    // As the tid is unique in this context, create only one
-    // writer/location per tid
-    const auto& location = registry_.emplace<otf2::definition::location>(
-        ByThreadMetricWriter(tid), intern(name),
-        registry_.get<otf2::definition::location_group>(ByProcess(pid)),
-        otf2::definition::location::location_type::metric);
-
-    return archive_(location);
-}
-
-otf2::writer::local& Trace::cpu_metric_writer(int cpuid)
-{
-    auto name = intern(fmt::format("metrics for cpu {}", cpuid));
-
-    const auto& location = registry_.emplace<otf2::definition::location>(
-        ByCpuMetricWriter(cpuid), name,
-        registry_.get<otf2::definition::location_group>(ByCpu(cpuid)),
-        otf2::definition::location::location_type::metric);
-    return archive_(location);
+    return archive_(intern_location);
 }
 
-otf2::writer::local& Trace::named_metric_writer(const std::string& name)
+// Normally we assume that names are unique, i.e. one name one location
+// But there are places where the name comes from an outside source (e.g. plugins)
+// where we can't assume uniqueness
+otf2::writer::local& Trace::non_unique_metric_writer(const std::string& name)
```

Comment:
I don't see the point of the comment here, as this is the only method that takes a name and returns a writer.
Also, I don't like the new name. Maybe `create` or `emplace` as a prefix works better.